File: src/CodexStatus.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Threading;
using System.Web.Script.Serialization;

// Only event metadata is persisted. Prompts, transcripts and tool arguments are discarded.
internal sealed class CodexRecord
{
    public string Session;
    public string Turn;
    public string Phase;
    public long Updated;
    public Dictionary<string, string> Tools = new Dictionary<string, string>();
    public Dictionary<string, bool> Waiting = new Dictionary<string, bool>();
}

internal sealed class CodexSnapshot
{
    public string Text = "Codex：等待连接";
    public string Phase = "disconnected";
    public int Active;
    public long Revision;
}

internal static class CodexStatus
{
    internal static string DirectoryPath = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "runtime", "codex-status");
    private static readonly JavaScriptSerializer Json = new JavaScriptSerializer { MaxJsonLength = 4194304 };

    private static string Value(Dictionary<string, object> input, string key)
    {
        object value;
        return input.TryGetValue(key, out value) && value is string ? (string)value : "";
    }

    internal static void Receive(string input)
    {
        var data = Json.Deserialize<Dictionary<string, object>>(input);
        string session = Value(data, "session_id"), turn = Value(data, "turn_id");
        string ev = Value(data, "hook_event_name"), tool = Value(data, "tool_name");
        if (session.Length == 0 || session.Length > 200 || turn.Length > 200) return;
        if (Array.IndexOf(new[] { "UserPromptSubmit", "PreToolUse", "PostToolUse", "PermissionRequest", "Stop", "Interrupt", "SessionEnd" }, ev) < 0) return;
        if (turn.Length == 0 && ev != "SessionEnd") return;
        string key;
        using (SHA256 hash = SHA256.Create())
            key = BitConverter.ToString(hash.ComputeHash(Encoding.UTF8.GetBytes(session))).Replace("-", "");
        Directory.CreateDirectory(DirectoryPath);
        string path = Path.Combine(DirectoryPath, key + ".json");
        using (Mutex gate = new Mutex(false, "Local\\GoldenMonkeyCodexStatusV1"))
        {
            bool held = false;
            try
            {
                try { held = gate.WaitOne(800); } catch (AbandonedMutexException) { held = true; }
                if (!held) return;
                CodexRecord record = File.Exists(path) ? Json.Deserialize<CodexRecord>(File.ReadAllText(path)) : null;
                if (record == null || (record.Turn != turn && ev == "UserPromptSubmit"))
                    record = new CodexRecord { Session = session, Turn = turn, Phase = "busy" };
                if (ev != "SessionEnd" && record.Turn != turn) return; // Late events cannot revive an old turn.
                string id = Value(data, "tool_use_id");
                if (id.Length == 0) id = tool.Length > 0 ? tool : "tool";
                if (ev == "UserPromptSubmit") record.Phase = "busy";
                else if (ev == "PreToolUse")
                {
                    record.Phase = "busy";
                    record.Tools[id] = tool == "apply_patch" || tool == "Edit" || tool == "Write" ? "edit" : "tool";
                }
                else if (ev == "PermissionRequest") { record.Phase = "busy"; record.Waiting[id] = true; }
                else if (ev == "PostToolUse") { record.Tools.Remove(id); record.Waiting.Remove(id); }
                else
                {
                    record.Phase = ev == "Stop" ? "ended" : ev == "Interrupt" ? "interrupted" : "closed";
                    record.Tools.Clear(); record.Waiting.Clear();
                }
                record.Updated = DateTime.UtcNow.Ticks;
                string temp = path + ".tmp";
                File.WriteAllText(temp, Json.Serialize(record), new UTF8Encoding(false));
                if (File.Exists(path)) File.Replace(temp, path, null); else File.Move(temp, path);
            }
            finally { if (held) gate.ReleaseMutex(); }
        }
    }

    internal static CodexSnapshot Read(DateTime now)
    {
        var result = new CodexSnapshot();
        if (!Directory.Exists(DirectoryPath)) return result;
        int waiting = 0, stale = 0;
        bool editing = false, tools = false;
        CodexRecord latest = null;
        foreach (string path in Directory.GetFiles(DirectoryPath, "*.json"))
        {
            CodexRecord record;
            try { record = Json.Deserialize<CodexRecord>(File.ReadAllText(path)); }
            catch (IOException) { continue; }
            catch (ArgumentException) { continue; }
            catch (InvalidOperationException) { continue; }
            if (record == null || record.Updated <= 0 || record.Tools == null || record.Waiting == null) continue;
            if (latest == null || record.Updated > latest.Updated) latest = record;
            result.Revision = Math.Max(result.Revision, record.Updated);
            if (record.Phase != "busy") continue;
            if (now.Ticks - record.Updated > TimeSpan.FromMinutes(30).Ticks) { stale++; continue; }
            result.Active++;
            waiting += record.Waiting.Count;
            editing |= record.Tools.ContainsValue("edit"); tools |= record.Tools.Count > 0;
        }
        if (latest == null) return result;
        if (waiting > 0) { result.Phase = "waiting"; result.Text = "Codex：需要确认"; }
        else if (result.Active > 0)
        {
            result.Phase = "busy";
            result.Text = editing ? "Codex：修改文件" : tools ? "Codex：执行工具" : "Codex：正在工作";
        }
        else if (stale > 0) { result.Phase = "unknown"; result.Text = "Codex：状态待更新"; }
        else
        {
            result.Phase = latest.Phase;
            bool recent = now.Ticks - latest.Updated < TimeSpan.FromSeconds(15).Ticks;
            result.Text = recent && latest.Phase == "ended" ? "Codex：本轮已结束"
                : recent && latest.Phase == "interrupted" ? "Codex：已中断" : "Codex：暂无活动";
        }
        if (result.Active > 1) result.Text += " ×" + result.Active;
        if (stale > 0 && result.Active > 0) result.Text += " ?";
        return result;
    }
}

File: macos/Sources/GoldenMonkeyShared/CodexStatus.swift
import Foundation
import CryptoKit

public struct CodexRecord: Codable {
    public var session: String
    public var turn: String
    public var phase: String
    public var updated: Date
    public var tools: [String: String]
    public var waiting: [String: Bool]

    enum CodingKeys: String, CodingKey {
        case session = "Session", turn = "Turn", phase = "Phase", updated = "Updated"
        case tools = "Tools", waiting = "Waiting"
    }
}

public struct CodexSnapshot {
    public var text = "Codex：等待连接"
    public var phase = "disconnected"
    public var active = 0
    public var revision = Date.distantPast
    public init() {}
}

public enum CodexStatus {
    public static let changedNotification = Notification.Name("com.sunyihao.goldenmonkeypet.codex-status-changed")
    private static let busyExpiry: TimeInterval = 5 * 60
    public static var directoryURL: URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return base.appendingPathComponent("GoldenMonkeyPet/codex-status", isDirectory: true)
    }

    public static func receive(_ input: Data, directory: URL? = nil) throws {
        guard let object = try JSONSerialization.jsonObject(with: input) as? [String: Any] else { return }
        let session = object["session_id"] as? String ?? ""
        let turn = object["turn_id"] as? String ?? ""
        let event = object["hook_event_name"] as? String ?? ""
        let tool = object["tool_name"] as? String ?? ""
        let accepted = ["UserPromptSubmit", "PreToolUse", "PostToolUse", "PermissionRequest", "Stop", "Interrupt", "SessionEnd"]
        guard !session.isEmpty, session.count <= 200, turn.count <= 200, accepted.contains(event) else { return }
        guard !turn.isEmpty || event == "SessionEnd" else { return }

        let root = directory ?? directoryURL
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        let digest = SHA256.hash(data: Data(session.utf8)).map { String(format: "%02X", $0) }.joined()
        let target = root.appendingPathComponent(digest + ".json")
        let lock = root.appendingPathComponent(digest + ".lock", isDirectory: true)
        guard acquireLock(lock) else { return }
        defer { try? FileManager.default.removeItem(at: lock) }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .millisecondsSince1970
        var record = (try? Data(contentsOf: target)).flatMap { try? decoder.decode(CodexRecord.self, from: $0) }
        if record == nil || (record?.turn != turn && event == "UserPromptSubmit") {
            record = CodexRecord(session: session, turn: turn, phase: "busy", updated: Date(), tools: [:], waiting: [:])
        }
        guard var value = record, event == "SessionEnd" || value.turn == turn else { return }
        let suppliedID = object["tool_use_id"] as? String ?? ""
        let id = suppliedID.isEmpty ? (tool.isEmpty ? "tool" : tool) : suppliedID
        switch event {
        case "UserPromptSubmit": value.phase = "busy"
        case "PreToolUse":
            value.phase = "busy"
            value.tools[id] = ["apply_patch", "Edit", "Write"].contains(tool) ? "edit" : "tool"
        case "PermissionRequest":
            value.phase = "busy"
            value.waiting[id] = true
        case "PostToolUse":
            value.tools.removeValue(forKey: id)
            value.waiting.removeValue(forKey: id)
        default:
            value.phase = event == "Stop" ? "ended" : event == "Interrupt" ? "interrupted" : "closed"
            value.tools.removeAll()
            value.waiting.removeAll()
        }
        value.updated = Date()
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .millisecondsSince1970
        let data = try encoder.encode(value)
        let temp = root.appendingPathComponent(digest + ".tmp")
        try data.write(to: temp, options: .atomic)
        if FileManager.default.fileExists(atPath: target.path) { try FileManager.default.removeItem(at: target) }
        try FileManager.default.moveItem(at: temp, to: target)
        DistributedNotificationCenter.default().postNotificationName(changedNotification, object: nil, userInfo: nil, deliverImmediately: true)
    }

    public static func read(now: Date = Date(), directory: URL? = nil) -> CodexSnapshot {
        var result = CodexSnapshot()
        let root = directory ?? directoryURL
        guard let files = try? FileManager.default.contentsOfDirectory(at: root, includingPropertiesForKeys: nil) else { return result }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .millisecondsSince1970
        var latest: CodexRecord?
        var waiting = 0, stale = 0
        var editing = false, tools = false
        for file in files where file.pathExtension == "json" {
            guard let data = try? Data(contentsOf: file), let record = try? decoder.decode(CodexRecord.self, from: data) else { continue }
            if latest == nil || record.updated > latest!.updated { latest = record }
            if record.updated > result.revision { result.revision = record.updated }
            guard record.phase == "busy" else { continue }
            if now.timeIntervalSince(record.updated) > busyExpiry { stale += 1; continue }
            result.active += 1
            waiting += record.waiting.count
            editing = editing || record.tools.values.contains("edit")
            tools = tools || !record.tools.isEmpty
        }
        guard let newest = latest else { return result }
        if waiting > 0 { result.phase = "waiting"; result.text = "Codex：需要确认" }
        else if result.active > 0 {
            result.phase = "busy"
            result.text = editing ? "Codex：修改文件" : tools ? "Codex：执行工具" : "Codex：正在工作"
        } else if stale > 0 { result.phase = "idle"; result.text = "Codex：暂无活动" }
        else {
            result.phase = newest.phase
            let recent = now.timeIntervalSince(newest.updated) < 15
            result.text = recent && newest.phase == "ended" ? "Codex：本轮已结束" : recent && newest.phase == "interrupted" ? "Codex：已中断" : "Codex：暂无活动"
        }
        if result.active > 1 { result.text += " ×\(result.active)" }
        if stale > 0 && result.active > 0 { result.text += " ?" }
        return result
    }

    private static func acquireLock(_ url: URL) -> Bool {
        for _ in 0..<40 {
            do { try FileManager.default.createDirectory(at: url, withIntermediateDirectories: false); return true }
            catch { Thread.sleep(forTimeInterval: 0.02) }
        }
        return false
    }
}
